clang/lib/Sema/SemaDeclCXX.cpp

```diff
@@ -11463,6 +11463,23 @@ void Sema::CheckExplicitObjectMemberFunction(Declarator &D,
     D.setInvalidType();
   }
 
+  // Handle the following case:
+  //
+  // struct S {
+  //   struct T {
+  //     int f(this T);
+  //   };
+  //
+  //   friend int T::f(this T); // Allow this.
+  //   friend int f(this S);    // But disallow this.
+  // };
+  if (D.getDeclSpec().isFriendSpecified() && D.getCXXScopeSpec().isEmpty()) {
```

Comment:
did you consider 

` friend int SomeNamespace::f(this T);` ?

Reply:
I’m candidly not sure anymore; I’ll add a test for that if there isn’t already one.

Reply:
Both of these cases are now diagnosed.